src/github.com/getlantern/flashlight/util/http.go

```diff
@@ -2,21 +2,162 @@ package util
 
 import (
 	"crypto/tls"
+	"errors"
 	"fmt"
 	"net"
 	"net/http"
 	"net/url"
 	"time"
 
+	"github.com/getlantern/fronted"
 	"github.com/getlantern/golog"
 	"github.com/getlantern/keyman"
 	"github.com/getlantern/waitforserver"
 )
 
+const (
+	defaultAddr = "127.0.0.1:8787"
+)
+
 var (
 	log = golog.LoggerFor("flashlight.util")
+
+	// This is for doing direct domain fronting if necessary. We store this as
+	// an instance variable because it caches TLS session configs.
+	direct  = fronted.NewDirect()
+	timeout = time.After(60 * time.Second)
 )
 
+// HTTPFetcher is a simple interface for types that are able to fetch data over HTTP.
+type HTTPFetcher interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// NewChainedAndFronted creates a new struct for accessing resources using chained
+// and direct fronted servers in parallel.
+func NewChainedAndFronted() *chainedAndFronted {
+	cf := &chainedAndFronted{}
+	dual := &dualFetcher{cf}
+
+	cf.fetcher = dual
+	cf.dual = dual
+	cf.chained = &chainedFetcher{}
+	return cf
+}
+
+// ChainedAndFronted fetches HTTP data in parallel using both chained and fronted
+// servers.
+type chainedAndFronted struct {
```

Comment:
I would create `chainedFetcher` and `dualFetcher` on the fly and assign to `fetcher` at that moment, as no extra cost involved. Three members with same type make me a bit confusion at first glance.